core/src/test/java/org/apache/accumulo/core/iterators/user/CombinerTest.java

```diff
@@ -786,4 +816,107 @@ public void testAdds() {
+      ret.put(new Key(combiner.getTopKey()), new Value(combiner.getTopValue()));
+      combiner.next();
+    }
+
+    return ret;
+  }
+
+  private void runDeleteHandlingTest(TreeMap<Key,Value> input, TreeMap<Key,Value> expected, DeleteHandlingAction dha, IteratorEnvironment env)
+      throws Exception {
+    runDeleteHandlingTest(input, expected, dha, env, null);
+  }
+
+  private void runDeleteHandlingTest(TreeMap<Key,Value> input, TreeMap<Key,Value> expected, DeleteHandlingAction dha, IteratorEnvironment env,
+      String expectedLog) throws Exception {
+    boolean deepCopy = expected == null;
+
+    StringWriter writer = new StringWriter();
+    WriterAppender appender = new WriterAppender(new PatternLayout("%p, %m%n"), writer);
+    Logger logger = Logger.getLogger(Combiner.class);
+    boolean additivity = logger.getAdditivity();
+    try {
+      logger.addAppender(appender);
+      logger.setAdditivity(false);
+
+      Combiner ai = new SummingCombiner();
+
+      IteratorSetting is = new IteratorSetting(1, SummingCombiner.class);
+      SummingCombiner.setEncodingType(is, LongCombiner.StringEncoder.class);
+      Combiner.setColumns(is, Collections.singletonList(new IteratorSetting.Column("cf001")));
+      if (dha != null) {
+        Combiner.setDeleteHandlingAction(is, dha);
+      }
+
+      ai.init(new SortedMapIterator(input), is.getOptions(), env);
+
+      if (deepCopy)
+        assertEquals(expected, readAll(ai.deepCopy(env)));
+      assertEquals(expected, readAll(ai));
+
+    } finally {
+      logger.removeAppender(appender);
+      logger.setAdditivity(additivity);
+    }
+
+    String logMsgs = writer.toString();
+    if (expectedLog == null) {
+      Assert.assertTrue(logMsgs, logMsgs.length() == 0);
+    } else {
+      logMsgs = logMsgs.replace('\n', ' ');
+      Assert.assertTrue(logMsgs, logMsgs.matches(expectedLog));
+    }
+  }
+
+  @Test
+  public void testDeleteHandling() throws Exception {
+    Encoder<Long> encoder = LongCombiner.STRING_ENCODER;
+
+    TreeMap<Key,Value> input = new TreeMap<Key,Value>();
+
+    IteratorEnvironment paritalMajcIe = new CombinerIteratorEnvironment(IteratorScope.majc, false);
+    IteratorEnvironment fullMajcIe = new CombinerIteratorEnvironment(IteratorScope.majc, true);
+
+    // keys that aggregate
+    nkv(input, 1, 1, 1, 1, false, 4l, encoder);
+    nkv(input, 1, 1, 1, 2, true, 0l, encoder);
+    nkv(input, 1, 1, 1, 3, false, 2l, encoder);
+    nkv(input, 1, 1, 1, 4, false, 9l, encoder);
+
+    TreeMap<Key,Value> expected = new TreeMap<Key,Value>();
+    nkv(expected, 1, 1, 1, 1, false, 4l, encoder);
+    nkv(expected, 1, 1, 1, 2, true, 0l, encoder);
+    nkv(expected, 1, 1, 1, 4, false, 11l, encoder);
+
+    try {
+      runDeleteHandlingTest(input, expected, DeleteHandlingAction.THROW_EXCEPTION, paritalMajcIe);
+      Assert.fail();
+    } catch (IllegalStateException ise) {
+      Assert.assertTrue(ise.getMessage().contains("Saw a delete during a partial compaction"));
```

Comment:
Try to include a meaningful error message if the assertion fails. We should be able to explain the exact reason the test failed just by looking at surefire output.